### ISSUE TITLE
numpy scalar in Dataframe not JSON serializable

### ISSUE THREAD

**Issue by boeddeker** [ellisonbg/altair#165]
Is it possible to allow numpy scalars in a dataframe?

The following code is not working, because a np.array is not json serializable:
```
df_numpy = pd.DataFrame([dict(
        x=np.array(1),
        y=np.array(2),
        )])
Chart(df_numpy).encode(x='x', y='y')
# TypeError: array(1) is not JSON serializable
```
Full Example: [
https://github.com/boeddeker/ipynb/blob/master/bug_altair_numpy.ipynb](
https://github.com/boeddeker/ipynb/blob/master/bug_altair_numpy.ipynb)

**Comment by jakevdp** [comment]
Thanks for the report!

I think the issue here is that you've created a dataframe containing object-types, rather than a dataframe containing numeric types. This is easier to see if you pass multi-valued arrays to your dataframe:

```python
df_numpy = pd.DataFrame([dict(
        x=np.array([1, 2, 3]),
        y=np.array([2, 3, 4, 5]),
        )])
>>> df_numpy
           x          y
0  [1, 2, 3]  [2, 3, 4, 5]
```

It's not entirely clear to me what you might expect Altair to do in this case: at each index, the column contains an array, which cannot be meaningfully serialized either as a numerical value or as a string.

Now it's true that the zero-dimensional array is perhaps a corner case in which we could specially convert the array-object-type to a single numerical scalar type, but that is a case we'd have to explicitly handle.

I think better would be to have the user give the data in a more standard form, e.g.

```python
df_numpy = pd.DataFrame(dict(
        x=np.array([1]),
        y=np.array([2]),
    ))
```

in which case the chart works as you would expect.

I'm curious how you came across this – do you think it's important for Altair to check for and correct such a corner-case for malformed inputs?

**Comment by jakevdp** [comment]
Thinking about it, one solution might be to preprocess the dataframes with something like this:
```python
for col in df_numpy.columns:
    try:
        df_numpy[col] = df_numpy[col].astype(float)
    except ValueError:
        pass
```
then you end up with a serializable array in this case. The problem is, though, that if you use this on an array of strings containing only digits, these digits will be converted to a numeric type. So before doing this, you'd have to individually check the Python type of each entry in any column of type object.

**Comment by boeddeker** [comment]
The problem is, that I use a library, which forces each value to be a numpy array. Because of that also scalars are numpy array's.

I currently try to replace matplotlib with Altair, but in my opinion this error should not appear, since other plotting library's can handle this. 

Since this example works
```
df_numpy = pd.DataFrame([dict(
        x=np.array(1).tolist(),
        y=np.array(2).tolist(),
        )])
```
I would implement something like this (maybe slow, because of the iteration over all elements)
```
for index, series in df_numpy.iterrows():
    for k, s in series.items():
        if isinstance(s, np.ndarray):
            df_numpy.loc[index][k] = s.tolist()
```
this would not break anything what currently is working.

An alternate is to use an JSONEncoder as described in http://stackoverflow.com/a/27050186 .

```
class MyEncoder(json.JSONEncoder):
    def default(self, obj):
        if isinstance(obj, numpy.integer):
            return int(obj)
        elif isinstance(obj, numpy.floating):
            return float(obj)
        elif isinstance(obj, numpy.ndarray):
            return obj.tolist()
        else:
            return super(MyEncoder, self).default(obj)

json.dumps(numpy.float32(1.2), cls=MyEncoder)
json.dumps(numpy.arange(12), cls=MyEncoder)
json.dumps({'a': numpy.int32(42)})
```

> I'm curious how you came across this – do you think it's important for Altair to check for and correct such a corner-case for malformed inputs?

I know that this is a special case, but the problem here is, that json can not handle numpy arrays and it seems, that the people form json don't want to solve this (http://bugs.python.org/issue18303, https://bugs.python.org/issue24313)

**Comment by jakevdp** [comment]
I see – yes I think we could add something along these lines to handle it. @ellisonbg – I'm curious about your opinion here: should this be handled via explicit conversion like we already do for some cases, or via a custom JSON encoder?

**Comment by jakevdp** [comment]
I pushed a fix in PR #167

**Comment by jakevdp** [comment]
@boeddeker, can you let me know if that solves the issue for you?
